operator/pkg/ingress/ingress_class.go

```diff
@@ -4,172 +4,201 @@
+		}
+
+		if event.Object != nil && event.Object.GetName() != ciliumIngressClassName {
+			event.Done(nil)
+
+			break
+		}
+
+		// !initialSync is passed as the value for enqueue in the below handlers
+		// because we should only pass events up to the ingress controller to
+		// allow it to react to changes in the state of the Cilium IngressClass.
+		// When performing an initial sync of the resource, we just need to
+		// learn if cilium is the default IngressClass and expose this
+		// information to the ingress controller via the IsDefault method.
+
+		switch event.Kind {
+		case resource.Upsert:
+			upsertHandler(event, !initialSync)
+		case resource.Delete:
+			deleteHandler(event, !initialSync)
+		default:
+			event.Done(nil)
+		}
+	case <-ctx.Done():
+		return ctx.Err()
+	}
 
-	ingressQueue workqueue.RateLimitingInterface
+	return nil
 }
 
-// type used to signal changes to the ingress controller queue
-type ciliumIngressClassUpdatedEvent struct {
-	ingressClass *slim_networkingv1.IngressClass
-}
+func (i *ingressClassManager) sync(ctx context.Context) error {
+	if i.synced.Load() {
+		return nil
+	}
 
-type ciliumIngressClassDeletedEvent struct {
-	ingressClass *slim_networkingv1.IngressClass
-}
+	log.Debug("Syncing IngressClass resources")
 
-func newIngressClassManager(clientset k8sClient.Clientset, ingressQueue workqueue.RateLimitingInterface, maxRetries int) (*ingressClassManager, error) {
-	manager := &ingressClassManager{
-		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
-		ingressQueue: ingressQueue,
-		maxRetries:   maxRetries,
+	if err := i.processEvents(ctx, i.handleSyncEvent, i.handleUpsertEvent, i.handleDeleteEvent, true); err != nil {
+		return fmt.Errorf("unable to sync IngressClass resources: %w", err)
 	}
 
-	manager.store, manager.informer = informer.NewInformer(
-		utils.ListerWatcherFromTyped[*slim_networkingv1.IngressClassList](clientset.Slim().NetworkingV1().IngressClasses()),
-		&slim_networkingv1.IngressClass{},
-		0,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    manager.handleAddIngressClass,
-			UpdateFunc: manager.handleUpdateIngressClass,
-			DeleteFunc: manager.handleDeleteIngressClass,
-		},
-		nil,
-	)
-
-	go manager.informer.Run(wait.NeverStop)
-	if !cache.WaitForCacheSync(wait.NeverStop, manager.informer.HasSynced) {
-		return manager, fmt.Errorf("unable to sync service")
-	}
-	log.WithField("existing-ingressclasses", manager.store.ListKeys()).Debug("ingress classes synced")
-	return manager, nil
+	i.synced.Store(true)
+
+	log.WithField(CiliumIngressClassIsDefault, i.isDefaultIngressClass.Load()).Info("IngressClass resources synced")
+
+	return nil
 }
 
 // Run kicks off the control loop
-func (i *ingressClassManager) Run() {
-	for i.processEvent() {
+func (i *ingressClassManager) Run(ctx context.Context) error {
+	// We should always be synced here, but just in case.
+	if !i.synced.Load() {
+		err := i.sync(ctx)
+		if err != nil {
+			return err
+		}
```

Comment:
My understanding is that, in the IngressController, we want to avoid handling Ingress events before syncing the IngressClass resource, otherwise we could end up in the scenario described in the commit message.

So, why not change the ingress controller Run so that:

- IngressClassManager Run loop starts first
- Ingress controller waits for IngressClass to be synced
- Ingress controller starts the ingressInformer

This should avoid the need for calling `sync` during IngressClassManager creation and repeat it here, too. Am I missing something?

Reply:
I don't think you're missing anything, I just think that the way I implemented this wasn't intuitive. In order for the Ingress controller to wait for IngressClass to be synced, we need to have the ingressClassManager's `sync` method be called. If `sync` returns an error, it's important that the error isn't dismissed, which is why I didn't want to call it in the ingress controller's `Run` method. That's why I put it in `newIngressClassManager`. I added a second potential call to `sync` in the ingressClassManger's `Run` method to make it clear that `sync` must be called before `Run`.

I made the `sync` method public and put a call to it in the ingress controller's constructor. I think this should clear things up. This way we have the following:

* Operator calls ingress.NewController
* ingressClassManager is constructed
* Inside ingress.NewController we call ingressClassManager.Sync
* If an error occurs, return it from ingress.NewController so it hits: 
https://github.com/cilium/cilium/blob/83cafbb9723fcde0093a412dc798ecb58f33067f/operator/cmd/root.go#L728-L731
* Call ingress controller's Run method which then calls the ingressClassManager's Run method (before the ingress informer is started).

WDYT?

Reply:
If stopping the operator is a requirement in case `Sync` errors out, I think this is surely better. :+1: 

But if we could just log the error in case of a failed sync (not dismissing it), then we can avoid starting two separate goroutines to execute the same events loop and move all in the `Run` controller method.
Using two goroutines forces us to write thread-safe code despite the events channel and it'll be easier to add racy code in case of changes.

Also, the only possible error returned by `Sync` seems to be related to context cancellation, that is triggered only during operator shutdown.

Reply:
I don't think I understand. The `Sync` method is not called in a separate goroutine. It's called in the main goroutine of the operator and is supposed to block the ingress controller until IngressClass resources are synced. After IngressClass resources are synced, the `Sync` method returns. When the ingress controller calls the ingress class manager's `Run` method, there will only be one goroutine processing events from the IngressClass events channel.

If we put everything in `Run`, then it makes it more difficult to have the ingress controller block until IngressClasses are synced. In order to avoid race conditions, IngressClasses must be synced before the Ingress informer is started. Putting everything in `Run` won't guarantee that, because it would be possible for the ingress controller to start processing Ingress resources before the ingress class manager has finished syncing.

Reply:
I just realized I might have misunderstood which `Run` method we were talking about. Anyways, I tried to make some adjustments to simplify the code and address your concerns. I renamed the `Sync` method to `WaitForSync`. It uses the `WaitForCacheSync` function from `k8s.io/client-go/tools/cache` to wait for a Sync event to be received. Now the `Run` method in the ingress controller looks like this:

```go
func (ic *Controller) Run(ctx context.Context) error {
	defer ic.queue.ShutDown()

	go ic.ingressClassManager.Run(ctx)

	// This should only return an error if the context is cancelled.
	if err := ic.ingressClassManager.WaitForSync(ctx); err != nil {
		return err
	}

	go ic.ingressInformer.Run(wait.NeverStop)
	if !cache.WaitForCacheSync(wait.NeverStop, ic.ingressInformer.HasSynced) {
		return fmt.Errorf("unable to wait for Ingress cache sync")
	}

	go ic.serviceManager.Run()
	go ic.secretManager.Run()

	for ic.processEvent() {
	}

	return nil
}
```